internal/querycoord/query_coord.go

```diff
@@ -425,5 +429,164 @@ func (qc *QueryCoord) watchHandoffSegmentLoop() {
+						segmentOffset[sourceNodeID]++
+						nodeID2MemUsage[sourceNodeID] -= uint64(segmentInfo.MemSize)
+						nodeID2MemUsage[dstNodeID] += uint64(segmentInfo.MemSize)
+						nodeID2MemUsageRate[sourceNodeID] = float64(nodeID2MemUsage[sourceNodeID]) / float64(nodeID2TotalMem[sourceNodeID])
+						nodeID2MemUsageRate[dstNodeID] = float64(nodeID2MemUsage[dstNodeID]) / float64(nodeID2TotalMem[dstNodeID])
+					} else {
+						// no enough memory on query nodes to balance, then notify proxy to stop insert
+						memoryInsufficient = true
+						break
+					}
+				} else {
+					// all query node's memoryUsageRate is less than 90%, and the max memUsageDiff is less than 30%
+					// this balance loop is done
+					break
+				}
+			}
+			if !memoryInsufficient {
+				for _, t := range loadBalanceTasks {
+					qc.scheduler.Enqueue(t)
+					log.Debug("loadBalanceSegmentLoop: enqueue a loadBalance task", zap.Any("task", t))
+					err = t.waitToFinish()
+					if err != nil {
+						// if failed, wait for next balance loop
+						// it may be that the collection/partition of the balanced segment has been released
+						// it also may be other abnormal errors
+						log.Error("loadBalanceSegmentLoop: balance task execute failed", zap.Any("task", t))
+					} else {
+						log.Debug("loadBalanceSegmentLoop: balance task execute success", zap.Any("task", t))
+					}
+				}
+				log.Debug("loadBalanceSegmentLoop: load balance Done in this loop", zap.Any("tasks", loadBalanceTasks))
+			} else {
+				// no enough memory on query nodes to balance, then notify proxy to stop insert
+				//TODO:: xige-16
+				log.Error("loadBalanceSegmentLoop: query node has insufficient memory, stop inserting data")
+			}
+		}
+	}
+}
+
+func generateLoadBalanceRequests(dstNodeID int64,
+	segmentInfo *querypb.SegmentInfo,
+	nodeID2MemUsage map[int64]uint64,
+	nodeID2TotalMem map[int64]uint64) (*querypb.LoadBalanceRequest, error) {
+	dstNodeMemUsageAfterBalance := nodeID2MemUsage[dstNodeID] + uint64(segmentInfo.MemSize)
+	dstNodeMemUsageRateAfterBalance := float64(dstNodeMemUsageAfterBalance) / float64(nodeID2TotalMem[dstNodeID])
+	// if memUsageRate of dstNode is greater than OverloadedMemoryThresholdPercentage after balance, than can't balance
+	if dstNodeMemUsageRateAfterBalance > Params.OverloadedMemoryThresholdPercentage {
+		return nil, fmt.Errorf("generateLoadBalanceRequests: query node %d don't have enough memory to balance", dstNodeID)
+	}
+	sourceNodeMemUsageAfterBalance := nodeID2MemUsage[segmentInfo.NodeID] - uint64(segmentInfo.MemSize)
+	sourceNodeMemUsageRateAfterBalance := float64(sourceNodeMemUsageAfterBalance) / float64(nodeID2TotalMem[segmentInfo.NodeID])
+	// if memUsageRateDIff of dstNode and sourceNode is greater than MemoryUsageMaxDifferencePercentage after balance, than can't balance
+	if dstNodeMemUsageRateAfterBalance-sourceNodeMemUsageRateAfterBalance > Params.MemoryUsageMaxDifferencePercentage {
```

Comment:
we shouldn't evaluate  whether we should do the move by MemoryUsageMaxDifferencePercentage
better evaluation would be 
dstNodeMemUsageRateAfterBalance-sourceNodeMemUsageRateAfterBalance < sourceNodeMemUsageRateBeforeBalance - dstNodeMemUsageRateBeforeBalance